### PR TITLE
Remove mocket

### PR DESCRIPTION
**Pull request by shadromani**

### CODE REVIEW

File: tests/webservice_test.py [hunk 1]
```diff
@@ -216,96 +211,86 @@ def test_300_error(self):
...
-        )
-        self.run_client(self.client.country("1.2.3.4"))
-        request = httpretty.last_request
+
+        request = None
+
+        def custom_handler(r):
+            nonlocal request
+            request = r
+            return ""
```
Comment (oschwald) [inline tests/webservice_test.py:279]: The custom handler is a bit confusing. I think we could eliminate it by using `headers` or `header_value_matcher` to match the expected headers on `expect_request`. I don't think we need to test the path again.

File: tests/webservice_test.py [hunk 2]
```diff
@@ -335,14 +317,13 @@ def test_city_ok(self):
...
         self.assertTrue(city.traits.is_anycast)
 
-    @httprettified
     def test_insights_ok(self):
-        httpretty.register_uri(
-            httpretty.GET,
-            self.base_uri + "insights/1.2.3.4",
-            body=json.dumps(self.insights),
+        self.httpserver.expect_request(
+            "/geoip/v2.1/insights/" + "1.2.3.4"
```
Comment (oschwald) [inline tests/webservice_test.py:322]:
```suggestion
            "/geoip/v2.1/insights/1.2.3.4"
```

File: tests/webservice_test.py [hunk 3]
```diff
@@ -335,14 +317,13 @@ def test_city_ok(self):
...
+        ).respond_with_json(
+            self.insights,
             status=200,
-            content_type=self._content_type("country"),
+            content_type=self._content_type("insights"),
```
Comment (oschwald) [inline tests/webservice_test.py:326]: :+1:

File: tests/webservice_test.py [hunk 4]
```diff
@@ -374,6 +355,7 @@ class TestClient(TestBaseClient):
     def setUp(self):
         self.client_class = Client
         self.client = Client(42, "abcdef123456")
+        self.client._base_uri = self.httpserver.url_for("/") + "geoip/v2.1"
```
Comment (oschwald) [inline tests/webservice_test.py:358]:
```suggestion
        self.client._base_uri = self.httpserver.url_for("/geoip/v2.1")
```

And below.

**Review by oschwald — CHANGES_REQUESTED** [review]
Nice! I had a few comments.

File: tests/webservice_test.py [hunk 5]
```diff
@@ -64,12 +60,13 @@ def _content_type(self, endpoint):
...
+    @pytest.fixture(autouse=True)
+    def setup_httpserver(self, httpserver: pytest_httpserver.HTTPServer):
+        self.httpserver = httpserver
+
     def test_country_ok(self):
-        httpretty.register_uri(
-            httpretty.GET,
-            self.base_uri + "country/1.2.3.4",
-            body=json.dumps(self.country),
+        self.httpserver.expect_request("/geoip/v2.1/country/1.2.3.4").respond_with_json(
```
[the remark this answers is not in the document]
Comment (shadromani) [reply tests/webservice_test.py:68]: Will you please elaborate how? The test has different responses depending on the specific ip and we can not return the default on all requests.

[the remark this answers is not in the document]
Comment (shadromani) [reply tests/webservice_test.py:68]: I see, I misread your comment.

File: tests/webservice_test.py [hunk 6]
```diff
@@ -270,39 +271,20 @@ def test_unknown_error(self):
...
-
-        def custom_handler(r):
-            nonlocal request
-            request = r
-            return ""
-
+        matcher = HeaderValueMatcher({
+            "Accept": "application/json",
+            "Authorization":"Basic NDI6YWJjZGVmMTIzNDU2",                
+            "User-Agent": lambda x: x.startswith("GeoIP2-Python-Client/"),})
```
Comment (oschwald) [inline tests/webservice_test.py]: I don't think this is the correct usage of `HeaderValueMatcher`. I think you still set `headers`, but you would want to define a custom matcher for "User-Agent".

[See the docs](https://pytest-httpserver.readthedocs.io/en/latest/howto.html#advanced-header-matching).

**Review by oschwald — CHANGES_REQUESTED** [review]

File: tests/webservice_test.py [hunk 7]
```diff
@@ -279,17 +279,27 @@ def test_unknown_error(self):
...
             "/geoip/v2.1/country/1.2.3.4",
             method="GET",
-            header_value_matcher=matcher,
+            header_value_matcher=HeaderValueMatcher(
+                {
+                    "User-Agent": user_agent_compare,
+                    "Accept": accept_compare,
+                    "Authorization": authorization_compare,
+                }
+            ),
```
Comment (oschwald) [inline tests/webservice_test.py:302]: I don't think this works either. If I change the values, the tests don't fail. I was expecting something like this:

```
        def user_agent_compare(actual: str, expected: str) -> bool:
            if actual is None:
                return False
            return actual.startswith(expected)

        self.httpserver.expect_request(
            "/geoip/v2.1/country/1.2.3.4",
            method="GET",
            headers={
                    "Accept": "application/json",
                    "Authorization": "Basic NDI6YWJjZGVmMTIzNDU2",
                    "User-Agent": "GeoIP2-Python-Client/"
            },
            header_value_matcher=HeaderValueMatcher(defaultdict(
                lambda: HeaderValueMatcher.default_header_value_matcher,
                {"User-Agent": user_agent_compare}
            )),
        ).respond_with_json(
            self.country,
            status=200,
            content_type=self._content_type("country"),
        )
```

Alternatively, you could modify the default matcher as directed in the docs.

**Review by oschwald — CHANGES_REQUESTED** [review]

**Review by oschwald — APPROVED** [review]

**Comment by deronnax** [comment]
I'm curious, what's the rational behind the switch from mocket to http-server?

**Comment by oschwald** [comment]
We unfortunately had quite a few problem with it breaking, e.g., on new or old Python versions, on new `aiohttp` versions, on particular OSes, etc. If you search the pull requests for "mocket", you will see a subset of those.

We decided to use `pytest-httpserver` as it implements an actual HTTP server, which should be less prone to these sorts of issues.

**Pull request merged**